### PR TITLE
React UI: Add Starting Screen

### PR DESCRIPTION
**Pull request by LeviHarrison**
This PR adds the `api/v1/status/walreplay` endpoint that tracks the WAL replay in real-time, and a page in the UI that is rendered when the server is starting up.

with-logo

And without the logo:

without-logo

### CODE REVIEW

File: web/ui/react-app/src/pages/starting/Starting.tsx
Comment (juliusv) [inline web/ui/react-app/src/pages/starting/Starting.tsx:25]: Hmm, it's a nice touch, but I'm not sure we should include a logo here, as we don't do that anywhere else. Of course, that's something we can think about separately some time on other pages, but let's not introduce it now just for a page that most people will never get to see.

File: web/ui/react-app/src/pages/starting/Starting.tsx
Comment (juliusv) [inline web/ui/react-app/src/pages/starting/Starting.tsx:26]: Could you share a screenshot in the PR of what this page looks like when the WAL is loading?

File: web/ui/react-app/src/pages/starting/Starting.tsx
Comment (juliusv) [inline web/ui/react-app/src/pages/starting/Starting.tsx:31]:
```suggestion
              Replaying WAL ({status?.read - status?.first}/{status?.last - status?.first})
```

File: web/api/v1/api.go
Comment (juliusv) [inline web/api/v1/api.go:1363]:
```suggestion
func (api *API) serveWALReplayStatus(w http.ResponseWriter, r *http.Request) {
```

File: tsdb/head.go
Comment (juliusv) [inline tsdb/head.go:57]: See https://github.com/golang/go/wiki/CodeReviewComments#initialisms for capitalization of initialisms like WAL.

**Review by juliusv — COMMENTED** [review]
Thanks, looking good in many ways already, but some comments :)

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1034]: Yeah, I was wondering what release this would fall into. I think it's probably `v2.26`, although I'm not exactly sure how your release cycles work and there's already an rc out.

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1060]: Yup, that seems to be right. Grammarly probably likes its hyphens a little bit too much 😄

File: web/ui/react-app/src/hooks/useFetch.ts
Comment (LeviHarrison) [reply web/ui/react-app/src/hooks/useFetch.ts:77]: My concern was with proxies, which still respond even if the endpoint isn't available, and those responses aren't handled as exceptions. I also now realize that `isResponding` (or something like it) should be set to false in the exception handler too, for instances that aren't behind proxies or other general network issues. Maybe just a blanket `isUnexpected` which can be used for non `200` and `503` status codes as well as network errors.

File: docs/querying/api.md
Comment (juliusv) [reply docs/querying/api.md:1034]: Since it's a new feature and not a bug fix, it'd go only into 2.27, since the 2.26 RC is already cut.

File: web/ui/react-app/src/hooks/useFetch.ts
Comment (juliusv) [reply web/ui/react-app/src/hooks/useFetch.ts:77]: Yep, something like that sounds fine I guess.

File: web/ui/react-app/src/pages/starting/Starting.tsx
Comment (LeviHarrison) [inline web/ui/react-app/src/pages/starting/Starting.tsx:28]: I just changed this so the loading bar won't un render once the replaying is done, and instead will turn green.

File: web/ui/react-app/src/pages/starting/Starting.tsx
Comment (LeviHarrison) [reply web/ui/react-app/src/pages/starting/Starting.tsx:26]: Just updated the comment with a (sadly low quality) GIF.

File: tsdb/head.go
Comment (LeviHarrison) [reply tsdb/head.go]: The method you're referring to relies on this data structure:

https://github.com/prometheus/prometheus/blob/bbc096cace886a5ca53cde72f56a36181e98d0b7/cmd/prometheus/main.go#L1056-L1062

Which is only populated with the `Head` once the instance is ready, and therefore only allows us to serve WAL replay status after it's already over. 

The current approach I'm leaning towards is having a variable in `head.go` like before, but having `api.go` call a function in `head.go` that reads it (after obtaining the lock), instead of having `api.go` read it directly like it currently does. This is the same way the variable containing the `readyStorage` value is handled.

Thoughts?

Comment (juliusv) [reply tsdb/head.go]: Ah, that makes things more complicated. Definitely we should solve this without globals though.

I wonder if we could restructure the whole readiness handling such that `readyStorage` holds the TSDB instance from the beginning (so `db` would never be `nil`), but instead its wrapper methods would determine readiness by calling the `WALReplayStatus()` method on the TSDB. Then I'd also extend the `TSDBAdminStats` interface at https://github.com/prometheus/prometheus/blob/9549a15c6ffd1e997ee15f6262e5a215a64d5661/web/api/v1/api.go#L150-L157 with a new `WALReplayStatus()` method and implement that in `readyStorage` as well. Just that that method would not have the usual readiness handling that `Stats()` etc. have, since it's already supposed to work before the TSDB is ready. Then you could use that method from the web API.

What do you think?

Comment (roidelapluie) [reply tsdb/head.go]: I think that this would be too much. I think we handle this differently without going to the global or changing the interfaces. This is very prometheus-tsdb specific.

**Comment by roidelapluie** [comment]
Thank you for this! This looks awesome.

I have a bug before the first segment is replayed:


I would like the progress bar to take less than the full width of the screen. I think this is too wide.

**Comment by roidelapluie** [comment]
If Prometheus is stopped while replaying the wall and comes back, the progressbar is not reset to 0 ( but it might be because of the bug in the previous comment )
wall

File: tsdb/head.go
Comment (LeviHarrison) [reply tsdb/head.go]: Oh sorry @roidelapluie I hadn't refreshed my page in a while before I submitted that comment.

I don't think that it would be too difficult, only requiring changes to a few functions/structers (`tsdb.Open`, `tsdb.open`, `*readyStorage.get`, `main.openDBWithMetrics`, `api.TSDBAdminStats` is all I can think of) and it shouldn't affect anything else beyond   `main.go`.

That being said, I would definitely be open to exploring other approaches.

**Comment by LeviHarrison** [comment]
I'm fairly certain I found the bug with the progress bar values, but I'll have to wait to push until I finish up with the WAL replay stuff on the server.

In the meantime, I changed the width:

Screenshot from 2021-03-29 16-02-31

How's this?

**Comment by roidelapluie** [comment]
I am wondering is a bar of the same size as the text would make sense.

**Comment by LeviHarrison** [comment]
Ok, here it is:

Screenshot from 2021-03-29 16-37-44

**Comment by LeviHarrison** [comment]
Also, on a whim, I noticed that when I `^C` (`SIGINT`) Prometheus in my terminal during the WAL replay, it will stop everything else but keep the replay going (ex: API calls stop responding). This is expected behavior, right?

**Comment by roidelapluie** [comment]
> Also, on a whim, I noticed that when I `^C` (`SIGINT`) Prometheus in my terminal during the WAL replay, it will stop everything else but keep the replay going (ex: API calls stop responding). This is expected behavior, right?

Mmmh I am not an expert here but at least your PR is not changing this behaviour from what I see.

**Comment by LeviHarrison** [comment]
Nope, I just noticed it because it prevents me from restarting the instance right after stopping it while testing.

**Comment by roidelapluie** [comment]
for the size, can we apply the css classes `col-2 mx-auto px-0` ?

**Comment by LeviHarrison** [comment]
Sure can!

Screenshot from 2021-03-29 17-40-57

File: tsdb/head.go
Comment (juliusv) [reply tsdb/head.go]: Thanks for the exploration & explanation! Yeah, generally, splitting the TSDB creation and initialization into two externally visible steps could be a decent way to go. Another option would be to pass a stats object into the `tsdb.Open()` call that is updated by the DB initialization, but can be read simultaneously by the caller.

Since we definitely need to touch the TSDB code in one way or another, I would just like to make sure that the TSDB maintainers @codesome and @bwplotka are aware of this. Maybe they also have a specific opinion / advice on how to most cleanly expose initialization progress.

**Comment by roidelapluie** [comment]
cc @codesome @bwplotka can you look at the tsdb questions in this issue?

File: tsdb/head.go
Comment (bwplotka) [reply tsdb/head.go]: Ack, looking on this part.

File: tsdb/head.go
Comment (bwplotka) [inline tsdb/head.go:57]: Is this global variable? I think that's might be not the best option. First of all, it will block people for running multiple Heads (e.g for test or any other purposes).

In fact we do this all the time in Thanos and Cortex projects.

File: tsdb/head.go
Comment (bwplotka) [inline tsdb/head.go:763]: I think there is a race here. What if API is reading from the variable in the moment it's writing? Unfortunately we might want to add some concurrency control (mutexes)

File: web/api/v1/api.go
Comment (bwplotka) [inline web/api/v1/api.go:1355]: This struct is already defined once in head, can we reuse it? 🤗

File: web/api/v1/api.go
Comment (bwplotka) [inline web/api/v1/api.go:1367]: As mentioned before, I would suggest actually passing this through API struct and add mutex (sync.Lock) to avoid undefined behaviours and even crashes.

**Review by bwplotka — CHANGES_REQUESTED** [review]
Thank you for this contribution! 💪🏽  I reviewed mostly the TSDB code and unfortunately, we need to add locking and move the status variable out of globals to make the API reliable. Let me know if you agree or I missed something.

Otherwise, it looks quite epic and will be super useful for everyone 🤗

File: tsdb/head.go
Comment (bwplotka) [reply tsdb/head.go]: I checked and ideally we get rid of the global.

@juliusv idea sounds quite accessible.

>  Another option would be to pass a stats object into the tsdb.Open() call that is updated by the DB initialization, but can be read simultaneously by the caller.

Otherwise, I would pass an additional function/interface to `API` that will have a direct connection to WAL/Head. Head can have `WalReplayStatus()` and we directly pass this method during initialization in Prometheus cmd/prometheus/main.go - no need to go through readiness storage 🤗

File: tsdb/head.go
Comment (LeviHarrison) [reply tsdb/head.go:57]: Yeah, in my working version of the codebase I had moved this into a field in the `Head` struct, added functions to modify the values that were concurrency safe, and also added the `WalReplayStats` (indirectly) to the `API` struct. I was just waiting on getting guidance on the head configuration thing so I could complete the work. Probably should have pushed as is - sorry for the trouble.

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1044]: `read` is the most recent segment to be replayed. So if we start at the 5th segment, then read it, the `read` value would be 5. In the front-end code, the `first` value is subtracted from the `read` value to generate the progress bar. I now realize this is flawed because the `read` value can be the same as the `first` value, showing no progress, even though that just means that only the first segment _has_ been read. The `read` value not being set to the starting value is what caused the bugs that @roidelapluie found. I think I'll probably need to rework this, maybe just making it the total number of segments read regardless of the ID, starting from 0, and then let the client (or server) figure the progress out.

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1051]: I like it! Should that be an addition, or completely replace the `first`, `read`, and last fields? I intentionally wrote this endpoint to expose more information than just the starting page needs in case anyone else would like to access the data, but I could support either direction.

File: web/api/v1/api.go
Comment (LeviHarrison) [reply web/api/v1/api.go:1355]: My issue here was the JSON mappings. They're required for the endpoint to work and I didn't know if we wanted them in structs not defined specifically for the API. Also, I wonder how the mutex would decode?. It might end up being private anyway, but I'm not sure.

**Comment by LeviHarrison** [comment]
Thanks so much for your input @bwplotka :) I just left a few responses, and I'll get to working as soon as I can!

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1042]: Since the WAL replay doesn't always start at the first segment, `first` is the position (starting from 1) of the first segment that needs to be read.

File: web/api/v1/api.go
Comment (LeviHarrison) [reply web/api/v1/api.go:1355]: Also, now we have the `WALReplayState`, which should probably be returned by the API in its string format, which requires a different struct definition.

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1044]: Furthermore, the reason that all of these fields are so specific (as to a single number counting from 0) is that I wanted this endpoint to provide the same data as the log lines do.

**Comment by LeviHarrison** [comment]
Ok, that should be all of it! I added the `DBStats` struct, which is passed into `Open` and furthermore into `NewHead` where it can be accessed during the WAL replay. I also changed the API response to be more succinct.

**Comment by LeviHarrison** [comment]
I'm pretty sure the current `test_tsdb_go115` failure is a flake.

**Comment by roidelapluie** [comment]
Could you rebase? Thanks

**Comment by LeviHarrison** [comment]
Gotcha!

**Comment by LeviHarrison** [comment]
I plan to add tests once the implementation of the head stats object has been agreed upon, so I can avoid changing them around a bunch after reviews.

**Comment by LeviHarrison** [comment]
It seems that this is going to miss 2.27.

File: web/api/v1/api.go
Comment (bwplotka) [reply web/api/v1/api.go:1356]: We can add mapping TBH to anything, but I think it's common for our trust to reproduce those, so not a big deal

File: tsdb/head.go
Comment (bwplotka) [reply tsdb/head.go]: ah, perfect, looing now!

File: docs/querying/api.md
Comment (bwplotka) [reply docs/querying/api.md]: Replace IMO (:

File: docs/querying/api.md
Comment (bwplotka) [reply docs/querying/api.md]: Yea, I think we don't need to be strict towards log line, will look how it looks now

File: cmd/prometheus/main.go
Comment (bwplotka) [inline cmd/prometheus/main.go:1095]: Do we need this method? Why not passing this through constructor always?

Comment (bwplotka) [reply cmd/prometheus/main.go:1095]: or even better, created through constructor?

File: docs/querying/api.md
Comment (bwplotka) [inline docs/querying/api.md:1058]:
```suggestion
    "total": 5,
```

File: docs/querying/api.md
Comment (bwplotka) [inline docs/querying/api.md:1059]: Do we need `read` if we have this progress?

File: docs/querying/api.md
Comment (bwplotka) [inline docs/querying/api.md:1060]: To ensure best API it's best to ensure the API is consistent and minimal. I still think it exposes too much information. As is there are multiple ways of getting the same information.

E.g to learn that the WAL replay is in progress I could:
* Check the percentage is higher than 0 and lower than 100
* I could check that `read` is higher than 0 and lower than `total`
* I could check the replay `state` == `in progress`

Having multiple ways of doing same thing for UI when talking to humans, sometimes makes sense, but for programmatic API it should be avoided. Every field adds extra cost. Adds and creates complexity, migration issues, documentation burden, support burden etc. 

Do you think we could aim for one minimal set of field that tells us everything?

I would really focus on literally one, two maybe 3 fields e.g:

```
status: "success", 
data:  {
  start_time: "<timestamp of replay start>"
  progress_percentage: 40,
  message: "replaying segment file 1", or doing X, doing Y, waiting for replay to start, parsing, finished - some human readable message

```

File: tsdb/db.go
Comment (bwplotka) [inline tsdb/db.go:592]: Why not just returning one?

File: tsdb/head.go
Comment (bwplotka) [inline tsdb/head.go:361]: Let's make sure we copy struct (so just return non pointer). We don't want to put pointer as it's still the same pointer to the same WALReplayStatus object so locking does not help ): 
```suggestion
func (s *WALReplayStatus) GetWALReplayStatus() WALReplayStatus {
```

File: tsdb/head.go
Comment (bwplotka) [inline tsdb/head.go:390]: Can head just give us the stats with some `GetStats()`? Why we need to compose it and create from outside? It's just same instance always and we don't need to be preffiled (we can make head being the owner of initialising this object)

**Review by bwplotka — COMMENTED** [review]
Looks good, just 2 important things one around HTTP API second around programmatic API.

Thanks!

File: docs/querying/api.md
Comment (roidelapluie) [reply docs/querying/api.md:1060]: message: "replaying segment file 1", or doing X, doing Y, waiting for replay to start, parsing, finished - some human readable message


Not really since we want to show a progressbar.

    "min": 210,
    "max": 271
    "current": 250,
    
   might be better

File: tsdb/db.go
Comment (LeviHarrison) [reply tsdb/db.go:592]: The whole problem in the first place is that the `open` function triggers the WAL replay, and we have to wait for that to finish until we get the return. So if we return the stats object, that means we can't get any of the WAL replay progress until it's actually completed, which defeats the whole point.

File: docs/querying/api.md
Comment (LeviHarrison) [reply docs/querying/api.md:1060]: Although `min`/`max`/`current` is a lot of fields, I think it's the best format because it provides machine-readable information about the magnitude of the replay not just for the starting screen but for other systems that might choose to process that data instead of just displaying it to a user (making the message less feasible). At the same time, I think we'll probably have to add in that state enum because checking those values alone isn't a reliable way to determine the status of the replay (if there are not segments, all three could be 0). I like the idea of `start_time` too, but maybe that's too many fields...

File: tsdb/head.go
Comment (LeviHarrison) [reply tsdb/head.go:390]: We don't get the DB object with the head back until after `Open` has returned, and thereby the WAL replay has been completed.

File: cmd/prometheus/main.go
Comment (LeviHarrison) [reply cmd/prometheus/main.go:1095]: If you mean adding it to the variable definition over here instead of using that function, I'm pretty sure that will work.

https://github.com/prometheus/prometheus/blob/9a19336b29060eb0414e80382b904777a002e91c/cmd/prometheus/main.go#L436-L437

Comment (bwplotka) [reply cmd/prometheus/main.go:1095]: Will work? Can we do this? Limiting number of methods is always helpful.

File: docs/querying/api.md
Comment (bwplotka) [reply docs/querying/api.md:1060]: So message is nice as it might tell what replay is focusing now, just nice to have. But yes not necessarily

Comment (LeviHarrison) [reply docs/querying/api.md:1060]: So should we add that on too, or just leave it at this?

Comment (bwplotka) [reply docs/querying/api.md:1060]: I would be against concrete number because:
* We may do things concurrently so suddenly number of segments will totally not aligned.
* We may do extra additional stages before, after or in between. 

I feel simple percentage allows us to decide on server side whatever percentage means and is the only thing that user needs as it's just loading bar, no?

Comment (roidelapluie) [reply docs/querying/api.md:1060]: This API is not covered by our stability guarantees. I like the fact that the UI shows what we have in the logs.

Comment (LeviHarrison) [reply docs/querying/api.md:1060]: The think the one thing a percentage is missing is the magnitude of segments that need to be loaded. 40% could be 4/10 or 400/1000; the ladder will take significantly longer. Maybe we just include the total number of segments that need to be loaded, and leave out the min and current. I think it will be nice for a user to be able to estimate how much longer it will take. On the other hand, I do see your point with the alignment and extra stages.

File: cmd/prometheus/main.go
Comment (LeviHarrison) [reply cmd/prometheus/main.go:1095]: Yup!

**Comment by LeviHarrison** [comment]
Sorry to anyone whose been pinged for code review, I accidentally messed up a rebase

**Comment by LeviHarrison** [comment]
I've changed the API response to `min`/`max`/`current`, but I suspect this is not over yet :) Again, I'll bring up _not_ passing the `DBStats` object to the main `tsdb.Open` method, and instead creating a separate `(*DBStats) Open(...)`. I've seen that `tsdb.Open` is widely used outside of Prometheus, and I just don't think it's worth it forcing everyone to add an extra `nil` for something they most likely won't use.

File: tsdb/head.go
Comment (LeviHarrison) [reply tsdb/head.go]: Govet complains about copying the lock value here (see failed tests)

**Comment by bwplotka** [comment]
Thanks a lot! Looking now

File: cmd/prometheus/main.go
Comment (bwplotka) [inline cmd/prometheus/main.go:1206]:
```suggestion
// WALReplayStatus implements the github.com/prometheus/prometheus/web/api/v1.TSDBStats interface.
```

File: docs/querying/api.md
Comment (bwplotka) [reply docs/querying/api.md:1060]: Cool, I think we find a good middleground (:

File: tsdb/db.go
Comment (bwplotka) [reply tsdb/db.go:593]: Ah you are right, thanks. Alternatively we could return unready DB, but not worth it (:

File: tsdb/head.go
Comment (bwplotka) [inline tsdb/head.go:335]:
```suggestion
	x := *s
```

Otherwise we still might have race when dereferencing on line 338 🙈 

But it's much better than it was!

**Review by bwplotka — APPROVED** [review]
I think this is ready to be merged, except for super minor nits! Thanks for this and for the patience! 💪🏽

File: tsdb/head.go
Comment (bwplotka) [reply tsdb/head.go:335]: Actually linter already mention this: 
image

Robots take my job away 😢

File: cmd/prometheus/main.go
Comment (LeviHarrison) [reply cmd/prometheus/main.go:1206]: The other function implementations don't though

File: tsdb/head.go
Comment (LeviHarrison) [reply tsdb/head.go:335]: It (still) complains about copying the lock value, is that a problem?

Comment (LeviHarrison) [reply tsdb/head.go]: Yeah, the linter is complaining about it now. Those darn robots...

**Comment by LeviHarrison** [comment]
Thanks so much! Before we merge, I think I should also probably add tests for the `DBStats` stuff, as well as the API route. And I'm still a little bit uneasy with the impact this may have on downstream projects:

> I'll bring up not passing the DBStats object to the main tsdb.Open method, and instead creating a separate (*DBStats) Open(...). I've seen that tsdb.Open is widely used outside of Prometheus, and I just don't think it's worth it forcing everyone to add an extra nil for something they most likely won't use.

File: cmd/prometheus/main.go
Comment (bwplotka) [reply cmd/prometheus/main.go:1206]: Nothing is perfect, feel free to ignore this then or fix the other too (:

**Review by bwplotka — APPROVED** [review]
I think extra param is not ideal, but we can live with that.

This PR LGTM (: Thanks for hard work here 💪🏽 

cc @juliusv @roidelapluie do you want to take another look?

File: tsdb/head.go
Comment (bwplotka) [inline tsdb/head.go:338]: This is still failing although there is nothing wrong given the WALReplayStatus does not  have pointers inside itself (that would not work):

```
GO111MODULE=on /go/bin/golangci-lint run --timeout 2m ./...
tsdb/head.go:335:7: copylocks: assignment copies lock value to x: github.com/prometheus/prometheus/tsdb.WALReplayStatus (govet)
	x := *s
	     ^
tsdb/head.go:338:9: copylocks: return copies lock value: github.com/prometheus/prometheus/tsdb.WALReplayStatus (govet)
	return x
```

I would suggest explicit copy then. Create new `WALReplayStatus` and copy fields manually. WDYT?

Comment (LeviHarrison) [reply tsdb/head.go:338]: Sure!

**Comment by bwplotka** [comment]
Let's merge and iterate if something is missing, it looks good for me! (:

**Comment by juliusv** [comment]
Thanks a bunch again!

Just testing this again (added a `time.Sleep()` in the Go code to simulate a slowly-starting storage). One thing I noticed: the starting screen is shown on initial page load, but since the `App` itself is only rendered once upon the initial load, it does not show the starting screen when I then further navigate to any other pages, like /targets:

notready-error

It's ok, but it would be even nicer if we could either prevent subsequently navigating to any specific pages (that don't work yet) or show a starting indicator on each of them, rather than an error. Further, the current behavior is that any page redirects to `/starting`, and when ready, redirects to `/`, instead of to the original page someone navigated to. A future improvement could indeed be to just have the starting spinner on all relevant pages and then stay on that page once the server is ready.

**Comment by LeviHarrison** [comment]
Makes sense, I'll add that when I can.

**Comment by roidelapluie** [comment]
Awesome!!

**Comment by LeviHarrison** [comment]
@juliusv #8909 fixes that.

**Pull request merged**
